src/jit/inline.cpp

```diff
@@ -169,14 +164,135 @@ InlineContext::InlineContext()
+// Return Value:
+//    InlineContext for use as the root context
+//
+// Notes:
+//    We leave inlCode as nullptr here (rather than the IL buffer
+//    address of the root method) to preserve existing behavior, which
+//    is to allow one recursive inline.
+
+InlineContext* InlineContext::newRoot(Compiler* compiler)
+{
+    InlineContext* rootContext = new (compiler, CMK_Inlining) InlineContext;
+
+#if defined(DEBUG)
+    rootContext->inlCallee = compiler->info.compMethodHnd;
+#endif
+
+    return rootContext;
+}
+
+//------------------------------------------------------------------------
+// newSuccess: construct an InlineContext for a successful inline
+// and link it into the context tree
+//
+// Arguments:
+//    compiler   - compiler doing the inlining
+//    stmt       - statement containing call being inlined
+//    inlineInfo - information about this inline
+//
+// Return Value:
+//    A new InlineContext for statements brought into the method by
+//    this inline.
+
+InlineContext* InlineContext::newSuccess(Compiler*   compiler,
+                                         InlineInfo* inlineInfo)
+{
+    InlineContext* calleeContext = new (compiler, CMK_Inlining) InlineContext;
+
+    GenTree*       stmt          = inlineInfo->iciStmt;
+    BYTE*          calleeIL      = inlineInfo->inlineCandidateInfo->methInfo.ILCode;
+    InlineContext* parentContext = stmt->gtStmt.gtInlineContext;
+
+    noway_assert(parentContext != nullptr);
+
+    calleeContext->inlCode = calleeIL;
+    calleeContext->inlParent = parentContext;
+    // Push on front here will put siblings in reverse lexical
+    // order which we undo in the dumper
+    calleeContext->inlSibling = parentContext->inlChild;
+    parentContext->inlChild = calleeContext;
+    calleeContext->inlChild = nullptr;
+    calleeContext->inlOffset = stmt->AsStmt()->gtStmtILoffsx;
+    calleeContext->inlObservation = inlineInfo->inlineResult->getObservation();
+#ifdef DEBUG
+    calleeContext->inlCallee = inlineInfo->fncHandle;
+    calleeContext->inlTreeID = inlineInfo->inlineResult->getCall()->gtTreeID;
+#endif
+
+    return calleeContext;
+}
+
 #ifdef DEBUG
 
 //------------------------------------------------------------------------
+// newFailure: construct an InlineContext for a failing inline
+// and link it into the context tree
+//
+// Arguments:
+//    compiler     - compiler doing the inlining
+//    stmt         - statement containing the attempted inline
+//    inlineResult - inlineResult for the attempt
+//
+// Return Value:
+//
+//    A new InlineContext for diagnostic purposes, or nullptr if
+//    the desired context could not be created.
+
+InlineContext* InlineContext::newFailure(Compiler*     compiler,
+                                         GenTree*      stmt,
+                                         InlineResult* inlineResult)
+{
+    // Check for a parent context first. We may insert new statements
+    // between the caller and callee that do not pick up either's
```

Comment:
It happens in `fgInlinePrependStatements` -- it can insert a cctor invocation, possibly more.